src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala

```diff
@@ -246,9 +245,73 @@ trait SubmissionComponent {
     }
 
     def loadSubmissionWorkflows(submissionId: UUID): ReadAction[Seq[Workflow]] = {
-      workflowQuery.findWorkflowsBySubmissionId(submissionId).result.flatMap(recs => DBIO.sequence(recs.map { rec =>
-        workflowQuery.get(rec.id) map(wf => wf.get)
-      }))
+      // get all workflows for this submission, with their entity and messages:
+      // WORKFLOW left join ENTITY left outer join WORKFLOW_MESSAGE
+      val workflows = for {
+        ((workflow, entity), message) <-
+          workflowQuery.filter(_.submissionId === submissionId) join
+            entityQuery on (_.workflowEntityId === _.id) joinLeft
+            workflowMessageQuery on (_._1.id === _.workflowId)
+      } yield (workflow, entity, message)
+
+      // get all input resolutions for all workflows in this submission
+      // we *could* do this in the same query as above, but since workflow:message and workflow:inputResolution
+      // are both one-to-many, this would cause a cartesian product
+      // WORKFLOW left outer join SUBMISSION_VALIDATION left outer join ATTRIBUTE
+      val inputResolutions = for {
+        (workflow, validationPair) <-
+        workflowQuery.filter(_.submissionId === submissionId) joinLeft {
+          submissionValidationQuery  joinLeft
+          attributeQuery on (_.valueId === _.id) } on (_.id === _._1.workflowId)
+      } yield (workflow, validationPair)
+      
+      inputResolutions.result.flatMap { resolutions =>
+        // first, prepare all the input resolutions
+        val resolutionsByWorkflowId = inputResolutions.result.map { resolutionRecs =>
```

Comment:
Oooooh, good catch, thank you. Refactoring this cleaned up a few more lines of code, in the latest commit.